### ISSUE TITLE
redirect_back not work as expected as previous_url

### ISSUE THREAD

**Issue by zw963** [luckyframework/lucky#1889]
**Describe the bug**

Please check following log:

1. It simple, render a form use `GET /universities/788/edit` from Index page (/universities)

```
web          |
web          | GET /universities/788/edit
web          |  ▸ Handled by Universities::Edit
web          | "Print in form"                    # => "Print in form"
web          | context.request.headers["Referer"] # => "http://127.0.0.1:3000/universities"
web          | previous_url(Universities::Index)  # => "http://127.0.0.1:3000/universities"
web          |  ▸ Rendered Universities::EditPage
web          |  ▸ Sent 200 OK (2.07ms)
```

Then submit the form use PUT action.

```
web          |
web          | PUT /universities/788
web          |  ▸ Handled by Universities::Update
web          | "Print in update action"   # => "Print in update action"
web          | request.headers["Referer"] # => "http://127.0.0.1:3000/universities/788/edit"
web          |  ▸ Sent 303 See Other (62.6ms)
web          |
web          | GET /universities/788/edit
web          |  ▸ Handled by Universities::Edit
web          | "Print in form"                    # => "Print in form"
web          | context.request.headers["Referer"] # => "http://127.0.0.1:3000/universities/788/edit"
web          | previous_url(Universities::Index)  # => "/universities"
web          |
web          |  ▸ Rendered Universities::EditPage
web          |  ▸ Sent 200 OK (2.29ms)
```

What i expected is to redirect to original url which is the entry for render the form use following code, But it not work

```crystal
class Universities::Update < BrowserAction
  include PageHelpers

  put "/universities/:university_id" do
    university = UniversityQuery.find(university_id)

    SaveUniversity.update(university, params) do |operation, updated_university|
      if operation.saved?
        pp! "Print in update action", request.headers["Referer"]
        flash.success = "修改成功"
        redirect_back fallback: Index, allow_external: true
      end
    else
      flash.failure = "出错了"
      html EditPage, operation: operation, university: updated_university
    end
  end
end
```

But, it not work as expected, as you can see the in above pp! log, the output of `request.headers["Referer"]` in the Update action output the `/universities/788/edit` instead of expected  `universities`(which print by the previous_url).

Thanks

**Expected behavior**

Use `redirect_back fallback: Index, allow_external: true` to redirect back to `/universities`

**Versions (please complete the following information):**
 - Lucky version (check in shard.lock): 1.2.0
 - Crystal version (`crystal --version`): 1.12.1
 - OS: Arch Linux

**Additional context**
I use htmx, i test with disable it, same issue.

**Comment by zw963** [comment]
Is there any info that needs to be update for clarify?

**Comment by matthewmcgarvey** [comment]
`redirect_back` is working as expected. When you went to the edit page from the universities index page, the index was the referrer. But when you submit the form, the referrer is the get request for the edit page which is why it sent you back to it.  I would recommend passing the url you want to send them back to as a field in the form if that's what you really want. Otherwise, just always redirect to the index

(Of course, I could be wrong).

**Comment by zw963** [comment]
This issue has been going on for some time, please let me try to reexplain it with more details.


1. When i visit `GET /universities/788/edit` from the index page ("http://127.0.0.1:3000/universities"), `I add some pp! code in the edit page`, as following, the output of `context.request.headers["Referer"]` and `previous_url(Universities::Index) ` in the form page both correct.

both output in the edit form page

web          | context.request.headers["Referer"] # => "http://127.0.0.1:3000/universities"
web          | previous_url(Universities::Index)  # => "http://127.0.0.1:3000/universities"

2. Let me submit the form

Then submit the form use PUT action, as you said in the reply, action print referer as edit page correctly

output in the action

request.headers["Referer"] # => "http://127.0.0.1:3000/universities/788/edit"

3.  We save the form successfully, the code is jump into the following `redirect_back` as following.

```crystal
      if operation.saved?
        pp! "Print in update action", request.headers["Referer"]
        flash.success = "修改成功"
>        redirect_back fallback: Index, allow_external: true
      else
        flash.failure = "出错了"
        html EditPage, operation: operation, university: updated_university
      end
```

4.  Then the following output is the pp! in the edit page. (redirect_back redirect page back to original edit page in this case)

```
web          | GET /universities/812/edit
web          |  ▸ Handled by Universities::Edit
web          | "Print in form" # => "Print in form"
web          | context.request.headers["Referer"] # => "http://127.0.0.1:3000/universities/812/edit"
web          | previous_url(Universities::Index) # => "/universities"
web          |  ▸ Rendered Universities::EditPage
web          |  ▸ Sent 200 OK (3.11ms)
```

5. so, the issue is, why `previous_url(Universities::Index)` return "/universities"? as the expected behavior said in your reply of  `redirect_back`,  previous_url should return edit page, right?

**Comment by jwoertink** [comment]
My guess is maybe a bug in `previous_url` ?

I'm not sure what this line is doing

https://github.com/luckyframework/lucky/blob/e06990316b0e64232d3717b7fc4cb665956b2eba/src/lucky/page_helpers/url_helpers.cr#L106

**Comment by zw963** [comment]
I meet same issue again when I start an new project use redirect_back.

I have a form, which is a user edit page,  like this: `form_for User::Update`,  the url is: `http://127.0.0.1:3000/user/edit`, after I submit the form, if use `redirect_back fallback: Docs::Index, allow_external: true` in the User::Update action.

Following is reproduce:

1. I am current in the url `docs/foo`
2. from `docs/foo` goto `/user/edit`
3. submit the form
4. What i really expected is: redirect to the `docs/foo`.
5.  But, the actually redirect to is: 
     - POST /user/update # <= correct
     - GET /user/edit  # <= incorrect,  it should be `docs/foo`, or fallback to Docs::Index.


If use `redirect Docs::Index`, it works as expected, so, it's must be redirect_back method set wrong url (it should be same as previous_url)

**Comment by zw963** [comment]
> I would recommend passing the url you want to send them back to as a field in the form if that's what you really want.

I consider this is the issue `redirect_back` really want to resolve.

**Comment by jwoertink** [comment]
Yeah, it sounds like a bug. I have no time to work on it, but I'm open to a PR.
